dnf/util.py

```diff
@@ -396,6 +396,64 @@ def _te_nevra(te):
     return nevra + te.V() + '-' + te.R() + '.' + te.A()
 
 
+def _log_rpm_trans_with_swdb(rpm_transaction, swdb_transaction):
+    logger.debug("Logging transaction elements")
+    for rpm_el in rpm_transaction:
+        tsi = rpm_el.Key()
+        tsi_state = None
+        if tsi is not None:
+            tsi_state = tsi.state
+        msg = "RPM element: '{}', Key(): '{}', Key state: '{}', Failed() '{}': ".format(
+            _te_nevra(rpm_el), tsi, tsi_state, rpm_el.Failed())
+        logger.debug(msg)
+    for tsi in swdb_transaction:
+        msg = "SWDB element: '{}', State: '{}', Action: '{}', From repo: '{}', Reason: '{}', " \
+              "Get reason: '{}'".format(str(tsi), tsi.state, tsi.action, tsi.from_repo, tsi.reason,
+                                        tsi.get_reason())
+        logger.debug(msg)
+
+
+def _sync_rpm_trans_with_swdb(rpm_transaction, swdb_transaction):
+    revers_actions = {libdnf.transaction.TransactionItemAction_DOWNGRADED,
+                      libdnf.transaction.TransactionItemAction_OBSOLETED,
+                      libdnf.transaction.TransactionItemAction_REMOVE,
+                      libdnf.transaction.TransactionItemAction_UPGRADED,
+                      libdnf.transaction.TransactionItemAction_REINSTALLED}
+    cached_tsi = [tsi for tsi in swdb_transaction]
+    el_not_found = False
+    error = False
+    for rpm_el in rpm_transaction:
+        te_nevra = _te_nevra(rpm_el)
+        tsi = rpm_el.Key()
+        if tsi is None or not hasattr(tsi, "pkg"):
+            for tsi_candidate in cached_tsi:
+                if tsi_candidate.state != libdnf.transaction.TransactionItemState_UNKNOWN:
+                    continue
+                if tsi_candidate.action not in revers_actions:
+                    continue
+                if str(tsi_candidate) == te_nevra:
+                    tsi = tsi_candidate
+                    break
+        if tsi is None or not hasattr(tsi, "pkg"):
+            logger.critical(_("TransactionItem not found for key: {}").format(te_nevra))
+            el_not_found = True
+            continue
+        if rpm_el.Failed():
+            tsi.state = libdnf.transaction.TransactionItemState_ERROR
+            error = True
+        else:
+            tsi.state = libdnf.transaction.TransactionItemState_DONE
+    for tsi in cached_tsi:
+        if tsi.state == libdnf.transaction.TransactionItemState_UNKNOWN:
+            logger.critical(_("TransactionSWDBItem not found for key: {}").format(str(tsi)))
+            el_not_found = True
+    if error:
+        logger.debug(_('Errors occurred during transaction.'))
+    if el_not_found:
+        _log_rpm_trans_with_swdb(rpm_transaction, cached_tsi)
+    return el_not_found
```

Comment:
The return value is never checked. Is it OK?

Reply:
I removed it, because after introduction of the logging functionality there is no use of the return value.